format/ComputeIR.fbs

```diff
@@ -0,0 +1,241 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+include "Schema.fbs";
+include "Message.fbs";
+
+namespace org.apache.arrow.flatbuf.computeir;
+
+/// An expression is one of
+/// - a Literal datum
+/// - a reference to a Field from a Relation
+/// - a call to a named function
+/// On evaluation, an Expression will have either array or scalar shape.
+union ExpressionImpl {
+  Literal, FieldRef, Call
+}
+
+table Expression {
+  impl: ExpressionImpl (required);
+}
+
+union Shape {
+  Array, Scalar
+}
+
+table Scalar {}
+
+table Array {
+  /// Number of slots.
+  length: long;
+}
+
+table Literal {
+  /// Shape of this literal.
+  shape: Shape (required);
+
+  /// The type of this literal.
+  type: Type (required);
+
+  /// Buffers containing `length` elements of arrow-formatted data.
+  /// If `length` is absent (this Literal is scalar), these buffers
+  /// are sized to accommodate a single element of arrow-formatted data.
+  /// XXX this can be optimized for trivial scalars later
+  buffers: [Buffer];
+}
+
+table FieldRef {
+  /// A sequence of field names to allow referencing potentially nested fields
+  path: [string];
+
+  /// For Expressions which might reference fields in multiple Relations,
+  /// this index may be provided to indicate which Relation's fields
+  /// `path` points into. For example in the case of a join,
+  /// 0 refers to the left relation and 1 to the right relation.
+  relation_index: int;
+
+  /// The type of data in the referenced Field.
+  type: Type;
+}
+
+table Call {
+  /// The name of the function whose invocation this Call represents.
+  function_name: string (required);
+
+  /// Parameters for `function_name`; content/format may be unique to each
+  /// value of `function_name`.
+  options: Buffer;
+
+  /// The arguments passed to `function_name`.
+  arguments: [Expression] (required);
+
+  /// The type of data which invoking `function_name` will return.
+  type: Type;
+}
+
+/// A relation is a set of rows with consistent schema.
+table Relation {
+  /// The namespaced name of this Relation.
```

Comment:
What is the namespace separator (`.`, `::`, ?)

Reply:
I had intended to use `::`, I'll clarify with examples